### PR TITLE
Port triangular_solve to structured kernel

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66401
* #62180
* #65606
* #64391
* #61536
* #63948
* #61858
* __->__ #61857

A few updates to internal code that allow marking triangular_solve as structured.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233

Differential Revision: [D31928687](https://our.internmc.facebook.com/intern/diff/D31928687)

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
There is going to be a conflict with https://github.com/pytorch/pytorch/pull/63567. That PR should be merged first.

**Comment by IvanYashchuk** [comment]
@ysiraichi, could you please take a look at this PR? This is my first attempt at porting to a function to a structured kernel.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1632,60 +1657,38 @@ static std::tuple<Tensor&, Tensor&> triangular_solve_out_info(
...
-  Tensor result = at::empty({0}, self.options());
-  Tensor clone_A = at::empty({0}, self.options());
-  Tensor infos = at::zeros({std::max<int64_t>(1, batchCount(self_broadcasted))}, self.options().dtype(kInt));
+  bool copy_needed = !result.transpose(-2, -1).is_contiguous();
+  copy_needed |= !clone_A.transpose(-2, -1).is_contiguous();
 
-  triangular_solve_out_info(result, clone_A, infos, A_broadcasted, self_broadcasted, upper, transpose, unitriangular);
+  if (copy_needed) {
+    Tensor result_tmp = at::empty({0}, self.options());
+    Tensor clone_A_tmp = at::empty({0}, self.options());
```
Comment (ysiraichi) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1683]: Shouldn't it be `A.options()` (not sure if it makes any difference, though)?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -1593,31 +1624,25 @@ The result of the computation is saved in-place in 'result' tensor,
...
   // the hierarchy of calls
-  TORCH_INTERNAL_ASSERT(input.dim() >= 2);
-  TORCH_INTERNAL_ASSERT(input.size(-2) == input.size(-1));
-
-  TORCH_INTERNAL_ASSERT(input.device() == other.device());
-  TORCH_INTERNAL_ASSERT(input.device() == result.device());
-  TORCH_INTERNAL_ASSERT(input.device() == clone_input.device());
-  TORCH_INTERNAL_ASSERT(input.device() == infos.device());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.dim() >= 2);
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.size(-2) == input.size(-1));
```
Comment (ysiraichi) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1637]: Are we doing this check in the META function?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -1593,31 +1624,25 @@ The result of the computation is saved in-place in 'result' tensor,
...
   // Error check with the actual error messages are done on the higher level of
   // the hierarchy of calls
-  TORCH_INTERNAL_ASSERT(input.dim() >= 2);
-  TORCH_INTERNAL_ASSERT(input.size(-2) == input.size(-1));
-
-  TORCH_INTERNAL_ASSERT(input.device() == other.device());
-  TORCH_INTERNAL_ASSERT(input.device() == result.device());
-  TORCH_INTERNAL_ASSERT(input.device() == clone_input.device());
-  TORCH_INTERNAL_ASSERT(input.device() == infos.device());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.dim() >= 2);
```
Comment (ysiraichi) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1636]: Nit: do we need these checks? I mean, is this function called by any others?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -203,6 +204,36 @@ extern "C" void strsm_(char *side, char *uplo, char *trans, char *diag, int *n,
...
+namespace meta {
+
+TORCH_META_FUNC(triangular_solve)(const Tensor& self, const Tensor& A, bool upper, bool transpose, bool unitriangular) {
+  TORCH_CHECK(self.dim() >= 2,
+           "torch.triangular_solve: Expected b to have at least 2 dimensions, but it has ", self.dim(), " dimensions instead");
+  TORCH_CHECK(A.dim() >= 2,
+           "torch.triangular_solve: Expected A to have at least 2 dimensions, but it has ", A.dim(), " dimensions instead");
+
+  Tensor self_broadcasted, A_broadcasted;
+  std::tie(self_broadcasted, A_broadcasted) = at::native::_linalg_broadcast_batch_dims(self, A, "triangular_solve");
```
Comment (ysiraichi) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:216]: This function could benefit from precomputed `Tensor` values (we wouldn't have to call `_linalg_broadcast_batch_dims` two times. But, AFAIU, they aren't yet supported, right @ezyang ?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -1593,31 +1624,25 @@ The result of the computation is saved in-place in 'result' tensor,
...
-  TORCH_INTERNAL_ASSERT(input.scalar_type() == result.scalar_type());
-  TORCH_INTERNAL_ASSERT(input.scalar_type() == clone_input.scalar_type());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.device() == other.device());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.device() == result.device());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.device() == clone_input.device());
 
-  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
-  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
-  TORCH_INTERNAL_ASSERT(infos.is_contiguous());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.scalar_type() == other.scalar_type());
```
Comment (ysiraichi) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1643]: Are we doing this check in the META function?

**Review by ysiraichi — COMMENTED** [review]
Hi @IvanYashchuk. Really nice PR! I've left a few minor comments. Let me know if something is unclear.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:216]: Precomputed values didn't land, but Tensors should never be precomputed!

Comment (ezyang) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:216]: You only make use of sizes() from the broadcasted elements. `_linalg_broadcast_batch_dims` should get refactored to have a variant that only returns sizes.

**Review by bdhirsh — APPROVED** [review]
LGTM!

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**